medmodels/data_synthesis/mtgan/modules/preprocessor.py

```diff
@@ -0,0 +1,595 @@
+
+        For instance, consider a single patient, 'p1', with admission dates as follows:
+        'p1': ['2021-01-01', '2021-01-15', '2021-01-25']
+
+        If the `time_interval_days` is 7 days, the method will compute time window as
+        the floor division of the difference in days from the first admission and the
+        `time_interval_days`. So, the resulting ABSOLUTE time windows will be:
+        'p1': [0, 2, 2]
+
+        Args:
+            medrecord (MedRecord): The MedRecord.
+            first_admission_attribute (MedRecordAttribute): The first admission
+                attribute.
+            time_interval_days (int): The time interval in days.
+            min_codes_per_window (int): The minimum number of codes per window.
+
+        Returns:
+            Tuple[MedRecord, PreprocessingAttributes]: The MedRecord with the relative
+                time windows assigned to edge. Also, the preprocessing attributes'
+                names in the MedRecord.
+
+        Raises:
+            ValueError: If the first admission attribute or the time attribute are not
+                datetime objects.
+
+        Note:
+            This method will remove the patient node if it has concepts in less than
+                two unique time windows.
+        """
+        absolute_time_window_attribute = self._get_attribute_name(
+            medrecord, "absolute_time_window"
+        )
+        time_window_attribute = self._get_attribute_name(medrecord, "time_window")
+        number_of_windows_attribute = self._get_attribute_name(
+            medrecord, "number_of_windows"
+        )
+
+        # TODO: need to perfect this with new queries.
+        for patient_index in sorted(medrecord.nodes_in_group(self.patients_group)):
+            first_admission = medrecord.node[patient_index, first_admission_attribute]
+            if not isinstance(first_admission, datetime.datetime):
+                raise ValueError(
+                    f"First admission attribute needs to be a datetime object, but got {type(first_admission)}"
+                )
+            for single_edge in medrecord.edges_connecting(
+                patient_index,
+                sorted(medrecord.nodes_in_group(self.concepts_group)),
+                directed=False,
+            ):
+                time = medrecord.edge[single_edge, self.time_attribute]
+                if not isinstance(time, datetime.datetime):
+                    raise ValueError(
+                        f"Time attribute needs to be a datetime object, but got {type(time)}"
+                    )
+                medrecord.edge[single_edge, absolute_time_window_attribute] = floor(
+                    (time - first_admission).days // time_interval_days
+                )
+                medrecord.edge[single_edge, concept_index_attribute] = (
+                    medrecord.select_nodes(
```

Comment:
So we're only keeping one concept per time window? How do we decide which one?

Reply:
No, we are keeping all of them. This is done inside a loop over all edges: Thus all concepts! Let me know if more clarification is needed.

Reply:
okay! What happens is that the return of all that is a list! So the [0] returns always the only element of that list (there can only be one element, since we are referring to one edge).